### PR TITLE
[vSphere] Support deploying one frozen VM, or a set of frozen VMs from OVF, then do ray up.

### PR DESCRIPTION
**Pull request by JingChen23**
### Bug fix
1. The default.yaml file was not built into the Python wheel, also not in the setup.py scirpt. This change added it.

#### 1. Support creating Ray nodes from a set of frozen VMs in a resource pool.

The motivation is when doing instant clone, the new VM must be on the same ESXi host with the parent VM. Previously we have only one frozen VM. The Ray nodes created from that frozen VM need to be relocated to other ESXi hosts by vSphere DRS. After this change, we can do round robin on the ESXi hosts to do instant clone to create the Ray nodes. We save the overhead of doing DRS.

#### 2. Support creating the frozen VM, or a set of frozen VMs from OVF template. 

This feature helps save some manual steps when the user has no existing frozen vm(s) but has an OVF template. Previously the user must manully login onto vSphere and deploy a frozen VM from the OVF first. Now we covered this fucntionality in ray up.

#### 3. Support powering on the frozen VM when the VM is at powered off status when doing ray up, we will wait the frozen VM is really "frozen", then do ray up.

Previously we have code logic to power on the frozen VM, but we will not wait it until it is frozen (usually need 2 mins or so).  This is a bug actually. In this change we add a function called "wait_until_frozen" to resolve this issue.

### Create one single frozen VM then do 'ray up'

The yaml snippet:
```
      frozen_vm:
        name: frozen-vm
        library_item: frozen-vm-1
        cluster: x77-cluster
        datastore: vsanDatastore
```
Verified that Ray up succeed.

### ray up with one existing frozen VM

The yaml snippet:
```
      frozen_vm:
        name: frozen-vm
        # library_item: frozen-vm-1
        # resource_pool: frozen-vms
        # cluster: x77-cluster
        # datastore: vsanDatastore
```
Verified that Ray up succeed.

### Creat a set of frozen VMs in a resource pool then do ray up, create ray nodes by round robin

The yaml snippet:
```
      frozen_vm:
        name: frozen-vm
        library_item: frozen-vm-item
        resource_pool: frozen-vms
        # cluster: vsphere-cluster
        datastore: vsanDatastore
```
Verified that Ray up succeed.
verified that the Ray nodes are spread on different ESXi hosts.

### Ray up on an existing resource pool of frozen VMs.

The yaml snippet:
```
      frozen_vm:
        #name: frozen-vm
        #library_item: frozen-vm-item
        resource_pool: frozen-vms
        # cluster: vsphere-cluster
        # datastore: vsanDatastore
```
Verified that Ray up succeed.
verified that the Ray nodes are spread on different ESXi hosts.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by architkulkarni** [comment]
@JingChen23 Just letting you know you can mark the PR as draft until it's ready for review! https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request#converting-a-pull-request-to-a-draft

File: python/ray/autoscaler/_private/vsphere/node_provider.py
Comment (JingChen23) [inline python/ray/autoscaler/_private/vsphere/node_provider.py:339]: We remove the functionalities of cloud init for now, because:
1. Our Packer project: https://github.com/vmware-ai-labs/vm-packer-for-ray, has already provided the functionality.
2. We have no doc to instruct the user about how to setup a user data for cloud init.

File: python/ray/autoscaler/_private/vsphere/scheduler.py
Comment (JingChen23) [inline python/ray/autoscaler/_private/vsphere/scheduler.py:9]: We may implement more schedulers in the future.

**Comment by architkulkarni** [comment]
@JingChen23 Looks like there are some potentially relevant test failures: https://buildkite.com/ray-project/premerge/builds/6411

**Review by architkulkarni — COMMENTED** [review]
The new API seems a bit complicated for someone unfamiliar with vSphere, but this might be unavoidable.  I think the following suggestions might mitigate this:

- Are there references in vSphere docs for the new terms "library item", "resource pool", "datastore" etc? It would be good to link to these.
- I think adding some example config snippets for `frozen_vm` in the docs would go a long way! The ones in the PR description might be a good starting point.
- There are a lot of constraints of the form "X must be specified, or if Y is specified Z must also be specified". Can we make sure these constraints are validated and can we add unit tests to make sure they fail fast with user friendly errors?

Other than this, looks good! Just minor comments.

The PR is a bit large, in the future it would be great to submit a series of smaller PRs.

**Comment by JingChen23** [comment]
> @JingChen23 Looks like there are some potentially relevant test failures: https://buildkite.com/ray-project/premerge/builds/6411

Thanks Archit, this is bacause I forgot to checkout the change on the UT file from our internal repo.

File: python/ray/autoscaler/_private/vsphere/pyvmomi_sdk_provider.py
Comment (JingChen23) [reply python/ray/autoscaler/_private/vsphere/pyvmomi_sdk_provider.py:17]: > If the session type is not UNVERIFIED, how should `context_obj` be initialized?

This is to be implemented by the next PRs, we already have that in our internal repo.

**Comment by JingChen23** [comment]
> The new API seems a bit complicated for someone unfamiliar with vSphere, but this might be unavoidable. I think the following suggestions might mitigate this:
> 
> * Are there references in vSphere docs for the new terms "library item", "resource pool", "datastore" etc? It would be good to link to these.
> * I think adding some example config snippets for `frozen_vm` in the docs would go a long way! The ones in the PR description might be a good starting point.
> * There are a lot of constraints of the form "X must be specified, or if Y is specified Z must also be specified". Can we make sure these constraints are validated and can we add unit tests to make sure they fail fast with user friendly errors?
> 
> Other than this, looks good! Just minor comments.
> 
> The PR is a bit large, in the future it would be great to submit a series of smaller PRs.

1. Reference added. The initial thought is that: the people who want to run Ray on vSphere should already have some basic knowledge of vSphere. This is a sellable product, it is a rare case that a person buy it but doesn't learn about it. 😄 
2. Example snipptes added.
3. We will add this in the next PR, basically my idea is to add a validator function to check the node config at the early stage, covering all the combinations. Then add a UT covers all the conbinations of the validator function.

**Review by architkulkarni — APPROVED** [review]
Looks good to me, as long as the unit tests are added in the next PR.

File: doc/source/cluster/vms/references/ray-cluster-configuration.rst
Comment (architkulkarni) [inline doc/source/cluster/vms/references/ray-cluster-configuration.rst:1229]: I'm not sure if raw urls are automatically converted to hyperlinks in RST, it would be good to check this in the doc build output, or just mimic how urls are written in other RST files

**Comment by architkulkarni** [comment]
Assigning @richardliaw as remaining codeowner

**Comment by JingChen23** [comment]
> Lint failed: https://buildkite.com/ray-project/premerge/builds/6504#018ad097-37b8-460b-b2c0-9fae441269bb/185-440
> 
> You can run setup_hooks.sh to install a pre-push hook that will automatically run lint and prevent this issue

Thank you! This bothers me for some time.

**Comment by JingChen23** [comment]
> Looks good to me, as long as the unit tests are added in the next PR.

I have added the unit test in this PR in the latest commit, yesterday I don't have time for that. Now that this PR had some issue and wasn't merged. I have plenty of time today so I paid the tech debt.

File: python/ray/autoscaler/_private/vsphere/config.py
Comment (architkulkarni) [inline python/ray/autoscaler/_private/vsphere/config.py:89]: I think `assert` should be replaced by `ValueError` with a descriptive error message, otherwise users will complain.

Comment (architkulkarni) [reply python/ray/autoscaler/_private/vsphere/config.py:89]: You can use `with pytest.raises(ValueError, match=...)` for the unit test. https://docs.pytest.org/en/7.1.x/how-to/assert.html#assertions-about-expected-exceptions

**Comment by architkulkarni** [comment]
Nice, thanks for adding the tests!

File: python/ray/autoscaler/_private/vsphere/config.py
Comment (architkulkarni) [reply python/ray/autoscaler/_private/vsphere/config.py:89]: Can be done in a followup PR

**Comment by JingChen23** [comment]
@architkulkarni @richardliaw Could you please help to merge this PR if the buildkites reports no isssues ralated to this?

**Comment by JingChen23** [comment]
The rst document

Screenshot 2023-09-27 at 10 28 25

**Review by edoakes — APPROVED** [review]

**Comment by architkulkarni** [comment]
Failed tests: `learning_tests_pendulum_ddppo`, `nested_action_spaces_ppo_torch`, `test_legacy_dataset_config` are all unrelated

**Pull request merged**
